### PR TITLE
Graphs

### PR DESCRIPTION
**Pull request by NathanielB123**

### CODE REVIEW

File: frontend/src/Utils/Utils.tsx [hunk 1]
```diff
@@ -167,7 +167,22 @@ export const mapSuccess = <T, U>(
...
+  m.length === 0 ? m : m[0].map((_, i) => m.map((row) => row[i]));
+
+export const all = <T,>(x: (T | null)[]): T[] | null =>
+  x.reduce<T[] | null>(
+    (p, c) => (c === null ? null : p?.concat([c]) ?? null),
+    [],
+  );
+
 // A mechine is busy if all gpus are in use
+// This utility really does not fit in this file...
```
Comment (sirjacobofglais) [inline frontend/src/Utils/Utils.tsx]: move it then

File: frontend/src/Components/DataTable.tsx [hunk 2]
```diff
@@ -13,34 +13,36 @@ import {
...
   Tr,
 } from "@chakra-ui/react";
-import { WorkStationGroup } from "../Data";
+import { GPUStats, WorkStationGroup } from "../Data";
 import { useState } from "react";
 import { useForceUpdate } from "framer-motion";
+import { keepIf } from "../Utils/Utils";
+
+// default to show group, machine_name, gpu_name, isFree, brand, and memory_total
+export const COLS: Record<string, boolean> = {
```
Comment (sirjacobofglais) [inline frontend/src/Components/DataTable.tsx]: The abstraction here isn't great: these bools are only relevant to this file

File: frontend/src/Components/Graph.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    return scale.ticks(numberOfTicksTarget).map((value) => ({
+      value,
+      offset: scale(value),
+    }));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [scale]);
+
+  return (
+    <>
+      {/* Main axis line */}
```
Comment (sirjacobofglais) [inline frontend/src/Components/Graph.tsx:104]: nit: multi-line comment on a single line

File: frontend/src/Components/WorkstationCardMinimal.tsx [hunk 4]
```diff
@@ -54,25 +63,42 @@ export const WorkstationCardMin = ({
...
+          <LinkBox>
+            <LinkOverlay
+              as={ReactRouterLink}
+              to={{ search: params.toString() }}
+            />
+            {gpus.map((s, i) => (
+              <Box key={i}>
+                <Heading size="md">{`${s.gpu_name} (${(
+                  s.memory_total / 1000
+                ).toFixed(0)} GB)`}</Heading>
```
Comment (sirjacobofglais) [inline frontend/src/Components/WorkstationCardMinimal.tsx:99]: I'm suprised prettier is happy with this

**Review by sirjacobofglais — APPROVED** [review]

File: frontend/src/Components/WorkstationCardMinimal.tsx (hunk 4 above)
Comment (NathanielB123) [reply frontend/src/Components/WorkstationCardMinimal.tsx:99]: unironically looks fine to me (I guess `Heading` tags could be on different lines to the `{}` content, but that would be 2 extra lines)

File: frontend/src/Components/Graph.tsx (hunk 3 above)
Comment (NathanielB123) [reply frontend/src/Components/Graph.tsx:104]: On one hand, I *did* just copy-paste this code from the above link, on the other, there is actually a reason for doing this - `//`-style comments would comment out the closing brace, so the comment would have to take up more lines. Happy to keep more consistent with the rest of the codebase and delete the comments if you want lmao

File: frontend/src/Utils/Utils.tsx (hunk 1 above)
Comment (NathanielB123) [reply frontend/src/Utils/Utils.tsx]: https://github.com/gpuctl/gpuctl/pull/221/commits/d6b6eda6d34337f03cc9a13196529f934ebfddbd

File: frontend/src/Components/DataTable.tsx (hunk 2 above)
Comment (NathanielB123) [reply frontend/src/Components/DataTable.tsx]: https://github.com/gpuctl/gpuctl/pull/221/commits/d6b6eda6d34337f03cc9a13196529f934ebfddbd

**Comment by NathanielB123** [comment]
There are a couple lingering issues ([#237](https://github.com/gpuctl/gpuctl/issues/237) and [#238](https://github.com/gpuctl/gpuctl/issues/238)) but this is already a pretty huge PR, so imma merge this now and we can address on those in follow-up PRs

**Pull request merged**